src/main/java/uk/gov/hmcts/ccd/domain/service/common/AccessControlService.java

```diff
@@ -120,40 +133,85 @@ public boolean canAccessCaseFieldsForUpsert(final JsonNode newData,
             final boolean noAccessGranted = getStream(newData)
                 .anyMatch(newFieldName -> {
                     if (existingData.has(newFieldName)) {
-                        return !valueDifferentAndHasUpdateAccess(newData,
-                            existingData,
-                            newFieldName,
-                            caseFieldDefinitions,
-                            userRoles);
+                        return !valueDifferentAndHasUpdateAccess(newData, existingData, newFieldName, caseFieldDefinitions, userRoles);
                     } else {
-                        return !hasCreateAccess(newFieldName,
-                            caseFieldDefinitions,
-                            userRoles);
+                        return !hasCaseFieldAccess(caseFieldDefinitions, userRoles, CAN_CREATE, newFieldName);
                     }
                 });
             return !noAccessGranted;
         }
         return true;
     }
 
-    public JsonNode filterCaseFieldsByAccess(final JsonNode caseFields,
-                                             final List<CaseField> caseFieldDefinitions,
-                                             final Set<String> userRoles,
-                                             final Predicate<AccessControlList> access) {
+    public JsonNode filterCaseFieldsByAccess(final JsonNode caseFields, final List<CaseField> caseFieldDefinitions,
+                                             final Set<String> userRoles, final Predicate<AccessControlList> access,
+                                             boolean isClassification) {
         JsonNode filteredCaseFields = JSON_NODE_FACTORY.objectNode();
-        if (caseFields != null) {
-            final Iterator<String> fieldNames = caseFields.fieldNames();
-            while (fieldNames.hasNext()) {
-                final String fieldName = fieldNames.next();
-                for (CaseField caseField : caseFieldDefinitions) {
-                    if (caseField.getId().equals(fieldName)
-                        && hasAccessControlList(userRoles, access, caseField.getAccessControlLists())) {
+        getStream(caseFields).forEach(
+            fieldName -> findCaseFieldAndVerifyHasAccess(fieldName, caseFieldDefinitions, userRoles, access)
+                .ifPresent(caseField -> {
+                    if (isEmpty(caseField.getComplexACLs())) {
                         ((ObjectNode) filteredCaseFields).set(fieldName, caseFields.get(fieldName));
+                    } else if (!isClassification) {
+                        ((ObjectNode) filteredCaseFields).set(fieldName, filterChildren(caseField, caseFields.get(fieldName), userRoles, access, isClassification));
```

Comment:
can we remove this cast and define filteredCaseFields as ObjectNode?

Reply:
👍